Entity/EventRepository.php

```diff
@@ -23,6 +23,51 @@ class EventRepository extends CommonRepository
 {
     use TimelineTrait;
 
+    /**
+     * Fetch the base event data from the database.
+     *
+     * @param                $contactSourceId
+     * @param                $eventType
+     * @param \DateTime|null $dateAdded
+     *
+     * @return array
+     */
+    public function getEventsByContactId($contactId, $eventType = null, \DateTime $dateAdded = null)
+    {
+        $q = $this->getEntityManager()->getConnection()->createQueryBuilder()
+            ->select([
+                'c.*',
+                'c.contact_id AS contactId',
+                'c.message',
+                'c.date_added AS timestamp',
+                'cs.name sourceName',
+            ])
+            ->from(MAUTIC_TABLE_PREFIX.'contactsource_events', 'c')
+        ->join('c', 'contactsource', 'cs', 'cs.id = c.contactsource_id');
+
+        $expr = $q->expr()->eq('c.contact_id', ':contactId');
+        $q->where($expr)
+            ->setParameter('contactId', (int) $contactId);
+
+        if ($dateAdded) {
+            $expr->add(
+                $q->expr()->gte('c.date_added', ':dateAdded')
+            );
+            $q->setParameter('dateAdded', $dateAdded);
+        }
+
+        if ($eventType) {
+            $expr->add(
+                $q->expr()->eq('c.type', ':type')
+            );
+            $q->setParameter('type', $eventType);
+        }
+
+        $results = $q->execute()->fetchAll();
```

Comment:
Let's add a limit to this query of 1. In the case of contactsource_events there should only be one row per contact_id (since it's a log record of the lead's original creation), so we can speed this up by adding a `$q->setMaxResults(1)` or fetchFirst() something like that.